execution/commitment/warmuper.go

```diff
@@ -0,0 +1,346 @@
+		numWorkers: numWorkers,
+		logPrefix:  logPrefix,
+		cache:      NewWarmupCache(),
+	}
+}
+
+// Start initializes and starts the warmup workers.
+func (w *Warmuper) Start() {
+	if w.started.Swap(true) {
+		return // Already started
+	}
+	if w.numWorkers <= 0 {
+		return
+	}
+
+	w.startTime = time.Now()
+	w.work = make(chan warmupWorkItem, 50_000)
+	w.g, w.ctx = errgroup.WithContext(w.ctx)
+
+	for i := 0; i < w.numWorkers; i++ {
+		w.g.Go(func() error {
+			trieCtx, cleanup := w.ctxFactory()
+			if cleanup != nil {
+				defer cleanup()
+			}
+
+			for item := range w.work {
+				select {
+				case <-w.ctx.Done():
+					return w.ctx.Err()
+				default:
+				}
+
+				w.warmupKey(trieCtx, item.hashedKey, item.startDepth)
+				w.keysProcessed.Add(1)
+			}
+			return nil
+		})
+	}
+}
+
+// warmupKey performs the actual warmup for a single key.
+func (w *Warmuper) warmupKey(trieCtx PatriciaContext, hashedKey []byte, startDepth int) {
+	depth := startDepth
+	for depth <= len(hashedKey) && depth <= w.maxDepth {
+		prefix := HexNibblesToCompactBytes(hashedKey[:depth])
+
+		// Check cache first
+		var branchData []byte
+		if entry, ok := w.cache.GetBranch(prefix); ok {
+			branchData = entry.Data
+		} else {
+			var step kv.Step
+			branchData, step, _ = trieCtx.Branch(prefix)
+			w.cache.SetBranch(prefix, branchData, step)
+		}
+
+		// Branch data format: 2-byte touch map + 2-byte bitmap + per-child data
+		if len(branchData) < 4 {
+			break
+		}
+
+		if depth >= len(hashedKey) {
+			break
+		}
+		nextNibble := int(hashedKey[depth])
+
+		// Extract and prefetch account/storage addresses
+		// Path nibble's cell will have stateHash cleared - always extract it
+		// Memoized siblings are skipped
+		cellAccounts, cellStorages := extractBranchCellAddresses(branchData, nextNibble)
+		for _, addr := range cellAccounts {
+			if _, ok := w.cache.GetAccount(addr); !ok {
+				update, _ := trieCtx.Account(addr)
+				w.cache.SetAccount(addr, update)
+			}
+		}
+		for _, addr := range cellStorages {
+			if _, ok := w.cache.GetStorage(addr); !ok {
+				update, _ := trieCtx.Storage(addr)
+				w.cache.SetStorage(addr, update)
+			}
+		}
+
+		branchData = branchData[2:] // skip touch map
+
+		bitmap := binary.BigEndian.Uint16(branchData[0:2])
+		childBit := uint16(1) << nextNibble
+
+		if bitmap&childBit == 0 {
+			break
+		}
+
+		// Find position of our child's data
+		pos := 2
+		for n := 0; n < nextNibble; n++ {
+			if bitmap&(uint16(1)<<n) != 0 {
+				if pos >= len(branchData) {
+					break
+				}
+				fieldBits := branchData[pos]
+				pos++
+				pos = skipCellFields(branchData, pos, fieldBits)
+			}
+		}
+
+		if pos >= len(branchData) {
+			break
+		}
+
+		fieldBits := branchData[pos]
+		pos++
+
+		// Check if child has extension
+		hasExtension := (fieldBits & 1) != 0
+		if hasExtension && pos < len(branchData) {
+			extLen, n := binary.Uvarint(branchData[pos:])
+			if n > 0 && extLen > 0 {
+				depth += int(extLen)
+				continue
+			}
+		}
+
+		depth++
```

Comment:
Blindly increasing depth like this will result in a lot of DB misses, since `hashedKey[:depth]` may not exist due to extension nodes. 

As an initial implementation it may be fine to keep it if it helps the overall result, but ultimately the set of prefixes to be prefetched is determined by the unfold. So in the future, we might as well keep everything touched in memory explicitly.